crates/corro-agent/src/agent/handlers.rs

```diff
@@ -401,6 +404,131 @@ fn processing_cost(change: &Changeset) -> usize {
     }
 }
 
+/// Handle incoming emptyset received during syncs
+///
+pub async fn handle_emptyset(
+    agent: Agent,
+    bookie: Bookie,
+    mut rx_emptysets: CorroReceiver<ChangeV1>,
+    mut tripwire: Tripwire,
+) {
+    // maybe bigger timeout?
+    let mut max_wait = tokio::time::interval(Duration::from_millis(
+        agent.config().perf.apply_queue_timeout as u64,
+    ));
+    let max_changes_chunk: usize = agent.config().perf.apply_queue_len;
+
+    let mut buf = HashMap::new();
+    let mut cost = 0;
+
+    loop {
+        let mut process = false;
+        tokio::select! {
+            maybe_change_src = rx_emptysets.recv() => match maybe_change_src {
+                Some(change) => {
+                    if let Changeset::EmptySet { versions, ts } = change.changeset {
+                        buf.entry(change.actor_id).or_insert(vec![]).push((versions.clone(), ts));
+                        cost += versions.len();
+                    } else {
+                        warn!("received non-emptyset changes in emptyset channel from {}", change.actor_id);
+                    }
+
+                    if cost >= max_changes_chunk && !buf.is_empty() {
+                        process = true;
+                    }
+                },
+                None => break,
+            },
+            _ = max_wait.tick() => {
+                process = true
+            },
+            _ = &mut tripwire => {
+                break;
+            }
+        }
+
+        if process {
+            let mut retain_keys = Vec::new();
+            for (actor, changes) in &buf {
+                match process_emptyset(agent.clone(), bookie.clone(), *actor, changes.clone()).await
```

Comment:
since you're awaiting without spawning, you can pass references instead of cloning.